src/zap.ts

```diff
@@ -999,18 +1018,759 @@ export class Zap {
+    const isTokenX = inputTokenMint.equals(tokenXMint);
+
+    const preInstructions: TransactionInstruction[] = [];
+    if (inputTokenMint.equals(NATIVE_MINT)) {
+      const { ataPubkey: userWrapSolAcc, ix: initializeWrapSolIx } =
+        await getOrCreateATAInstruction(
+          this.connection,
+          NATIVE_MINT,
+          user,
+          user,
+          false,
+          TOKEN_PROGRAM_ID
+        );
+      const wrapSOL = wrapSOLInstruction(
+        user,
+        userWrapSolAcc,
+        BigInt(amountIn.toString())
+      );
+      initializeWrapSolIx && preInstructions.push(initializeWrapSolIx);
+      preInstructions.push(...wrapSOL);
+    }
+
+    const swapTransactions: Transaction[] = [];
+    let maxTransferAmount: BN;
+
+    if (
+      directSwapEstimate.swapDirection !== "noSwap" &&
+      directSwapEstimate.quote
+    ) {
+      const swapQuote = directSwapEstimate.quote;
+      if (swapQuote.route === "jupiter") {
+        const { transaction: swapTx } = await buildJupiterSwapTransaction(
+          user,
+          directSwapEstimate.swapDirection === "xToY" ? tokenXMint : tokenYMint,
+          directSwapEstimate.swapDirection === "xToY" ? tokenYMint : tokenXMint,
+          directSwapEstimate.swapAmount,
+          maxAccounts,
+          slippageBps
+        );
+        swapTransactions.push(swapTx);
+      } else {
+        const swapForY = directSwapEstimate.swapDirection === "xToY";
+        const binArrays = await dlmm.getBinArrayForSwap(swapForY);
+        const swapTx = await dlmm.swap({
+          inToken: swapForY ? tokenXMint : tokenYMint,
+          outToken: swapForY ? tokenYMint : tokenXMint,
+          inAmount: directSwapEstimate.swapAmount,
+          minOutAmount: directSwapEstimate.expectedOutput,
+          lbPair: lbPair,
+          user,
+          binArraysPubkey: binArrays.map((item) => item.publicKey),
+        });
+
+        // Remove close account instructions if swapping involves native SOL
+        // The wrapped SOL account will be needed by the zap instruction
+        if (tokenXMint.equals(NATIVE_MINT) || tokenYMint.equals(NATIVE_MINT)) {
+          swapTx.instructions = swapTx.instructions.filter((ix) => {
+            // Filter out CloseAccount instructions (discriminator = 9 for SPL Token)
+            if (ix.programId.equals(TOKEN_PROGRAM_ID)) {
+              return ix.data[0] !== 9;
+            }
+            return true;
+          });
+        }
+        swapTransactions.push(swapTx);
+      }
+      maxTransferAmount = getExtendMaxAmountTransfer(
+        directSwapEstimate.expectedOutput.toString(),
+        maxTransferAmountExtendPercentage
+      );
+    } else {
+      maxTransferAmount = new BN(0);
+    }
+
+    const amount = isTokenX
+      ? directSwapEstimate.postSwapX
+      : directSwapEstimate.postSwapY;
+
+    const cleanUpInstructions: TransactionInstruction[] = [];
+    if (
+      inputTokenMint.equals(NATIVE_MINT) ||
```

Comment:
don't need to check inputMint is NATIVE_MINT here.
because we already have invariant check input_mint always is tokenX or tokenY